pkg/tsdb/cloudwatch/get_dimension_values_for_wildcards_test.go

```diff
@@ -0,0 +1,78 @@
+		assert.Nil(t, err)
+		assert.Len(t, queries, 1)
+		assert.NotNil(t, queries[0].Dimensions["Test_DimensionName1"], 1)
+		assert.Equal(t, []string{"Value1"}, queries[0].Dimensions["Test_DimensionName1"])
+	})
+
+	t.Run("Should not change exact dimension value", func(t *testing.T) {
+		query := getBaseQuery()
+		query.MetricName = "Test_MetricName1"
+		query.Dimensions = map[string][]string{"Test_DimensionName1": {"*"}}
+		queries, err := executor.getDimensionValuesForWildcards(context.Background(), nil, "us-east-1", []*models.CloudWatchQuery{query})
+		assert.Nil(t, err)
+		assert.Len(t, queries, 1)
+		assert.NotNil(t, queries[0].Dimensions["Test_DimensionName1"])
+		assert.Equal(t, []string{"*"}, queries[0].Dimensions["Test_DimensionName1"])
+	})
+
+	t.Run("Should change wildcard dimension value", func(t *testing.T) {
+		query := getBaseQuery()
+		query.MetricName = "Test_MetricName1"
+		query.Dimensions = map[string][]string{"Test_DimensionName1": {"*"}}
+		query.MatchExact = false
+		api := &mocks.MetricsAPI{Metrics: []*cloudwatch.Metric{
+			{MetricName: utils.Pointer("Test_MetricName1"), Dimensions: []*cloudwatch.Dimension{{Name: utils.Pointer("Test_DimensionName1"), Value: utils.Pointer("Value1")}, {Name: utils.Pointer("Test_DimensionName2"), Value: utils.Pointer("Value2")}}},
+			{MetricName: utils.Pointer("Test_MetricName2"), Dimensions: []*cloudwatch.Dimension{{Name: utils.Pointer("Test_DimensionName1"), Value: utils.Pointer("Value3")}}},
+			{MetricName: utils.Pointer("Test_MetricName3"), Dimensions: []*cloudwatch.Dimension{{Name: utils.Pointer("Test_DimensionName1"), Value: utils.Pointer("Value4")}}},
+			{MetricName: utils.Pointer("Test_MetricName4"), Dimensions: []*cloudwatch.Dimension{{Name: utils.Pointer("Test_DimensionName1"), Value: utils.Pointer("Value2")}}},
+		}}
+		api.On("ListMetricsPages").Return(nil)
+		queries, err := executor.getDimensionValuesForWildcards(context.Background(), api, "us-east-1", []*models.CloudWatchQuery{query})
+		assert.Nil(t, err)
+		assert.Len(t, queries, 1)
+		assert.Equal(t, map[string][]string{"Test_DimensionName1": {"Value1", "Value2", "Value3", "Value4"}}, queries[0].Dimensions)
+		api.AssertExpectations(t)
+	})
+
+	t.Run("Should use cache for previously fetched value", func(t *testing.T) {
+		query := getBaseQuery()
+		query.MetricName = "Test_MetricName2"
+		query.Dimensions = map[string][]string{"Test_DimensionName2": {"*"}}
+		query.MatchExact = false
+		api := &mocks.MetricsAPI{Metrics: []*cloudwatch.Metric{
+			{MetricName: utils.Pointer("Test_MetricName2"), Dimensions: []*cloudwatch.Dimension{{Name: utils.Pointer("Test_DimensionName2"), Value: utils.Pointer("Value1")}}},
+		}}
+		api.On("ListMetricsPages").Return(nil)
+		_, err := executor.getDimensionValuesForWildcards(context.Background(), api, "us-east-1", []*models.CloudWatchQuery{query})
+		assert.Nil(t, err)
+
```

Comment:
Unfortunately this test does not assert quite what we think.
The `query` input is mutated by the results from the first call to `ListMetricsPages`, so what happens during the second call is that the dimension is no longer a wildcard. The loop `continue`s. 
The assertions still pass, however, since the result is that `ListMetricsPages` doesn't get called again.

I'm aware I'm stepping back here, but would it be possible to not mutate the `queries []*models.CloudWatchQuery` in `getDimensionValuesForWildcards`? We are already using a function signature which returns the result type we want. I think it may be misleading to mutate the input as well as provide the output in the function signature. Let's make some copies internally and return that, what do you think?

Reply:
Changed to not mutate the queries.